File: qqbot-module-spi/src/main/java/com/mieai/qqbot/module/spi/ModuleContext.java
package com.mieai.qqbot.module.spi;

import com.mieai.qqbot.module.api.ModuleServiceKey;
import com.mieai.qqbot.module.api.ModuleServiceRegistration;
import java.util.Optional;

/** Lifecycle-scoped service exchange available to one framework module. */
public interface ModuleContext {
    String moduleId();
    <T> ModuleServiceRegistration publish(ModuleServiceKey<T> key, T service);
    <T> Optional<T> find(ModuleServiceKey<T> key);

    default <T> T require(ModuleServiceKey<T> key) {
        return find(key).orElseThrow(() -> new IllegalStateException(
                "Required module service is unavailable: " + key.name()));
    }
}

File: qqbot-module-spi/src/main/kotlin/com/mieai/qqbot/module/spi/ModuleContext.kt
package com.mieai.qqbot.module.spi

import com.mieai.qqbot.module.api.ModuleServiceKey
import com.mieai.qqbot.module.api.ModuleServiceRegistration

/** Lifecycle-scoped service exchange available to one framework module. */
interface ModuleContext {
    val moduleId: String

    fun <T : Any> publish(key: ModuleServiceKey<T>, service: T): ModuleServiceRegistration

    fun <T : Any> find(key: ModuleServiceKey<T>): T?

    fun <T : Any> require(key: ModuleServiceKey<T>): T = find(key)
        ?: throw IllegalStateException("Required module service is unavailable: ${key.name}")
}
